### DISCUSSION TITLE
How to limit number of parallel workflows?

### DISCUSSION THREAD

**Discussion by vitalyk-multinarity** [community/community#30504] [Actions]
In my workflow, I'm using large GCP VM instances (many CPUs + GPU) for self-hosted runners.  Due to GCP limits, I can launch up to 5 VM instances. 
I launch my runners dynamically, i.e. my workflow is composed of 3 jobs:
1) launch a self-hosted runner
2) run heavy ML job
3) delete runner

So I'd like to limit the number of parallel executions of my workflow to 5; if the next job is triggered, it should be queued. 
It seems to be a trivial feature for any CI platform, but I couldn't see how to implement this in Github Actions.
Any ideas?
Thanks, Vitaly

PS: I hope now I explained myself better than last time https://github.com/orgs/community/discussions/30270.

**Comment by airtower-luna** [comment]
You shouldn't need to do anything, jobs that can't get a free runner immediately are automatically queued. The [is a limit on how long a job can stay queued](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#usage-limits), though.

**Reply by vitalyk-multinarity** [reply]
My fault - I didn't mention that I launch my runners dynamically, i.e. my workflow is  composed from 3 jobs:
1) launch self-hosted runner
2) run heavy ML job
3) delete runner

**Comment by naikrovek** [comment]
i don’t understand.  you can only launch 5 but you are launching more?  don’t launch more if you already have 5 going.

what are you using to accept the webhook payloads and spawn virtual machines?

**Reply by vitalyk-multinarity** [reply]
@naikrovek - on my 1st step I'm using 'gcloud' CLI for launching a new VM for GHA self-hosted runner. If GCP doesn't allow me to launch a new VM, my workflow fails. I don't want to build my DIY orchestration to manage a number of parallel workflows, I'm looking for a basic throttling feature that exists in Jenkins and other CI platforms.

**Reply by naikrovek** [reply]
so you are launching VMs when you receive the "queued" webhook for an Actions job.

it sounds like you will need some logic to make a decision when a webhook payload is received rather than always attempting to launch a VM.

this is how GitHub actions works.  the job will not progress to the "in_progress" state until the runner software itself has taken the job.  but also, as you have discovered, the "queued" event is never automatically re-sent.

you will need to keep track of the status of the virtual machines, and wait until one of your five slots are available before you attempt to launch that virtual machine.  GitHub actions does not provide this directly because it is the job of the orchestrator to handle this, and not GitHub itself.  that is by design.

**Reply by vitalyk-multinarity** [reply]
@naikrovek - thank you for clarification.

> GitHub actions does not provide this directly because it is the job of the orchestrator to handle this, and not GitHub itself. that is by design.

I'm not sure I understand this logic. IMHO it should the job of CI platform. If we already have "concurrency" in GHA, currently it's hardcoded to "1", why GHA cannot allow custom value to "concurrency"?

**Reply by naikrovek** [reply]
there is no such setting for GHA, and it is not hardcoded to 1.  each job in a workflow will happen in parallel by default, unless you specify in the workflow that one job relies on the output of another or that one job should wait for another.  

in fact, you can specify 255 jobs in a single workflow file, and when you trigger the workflow, 255 jobs will run in parallel, if there are that many runners available to do the work.  Your webhook receiver will get 255 "queued" events at once, in this scenario.

GHA itself has no concept of throttling as you describe.  You set the number of runners that you want to run at once by having that many runners online and idle, waiting for work, or you handle such things in your orchestration tool, which would keep track of how many runners you currently have running, and wait until that number drops below some maximum that you declare before creating another.  Custom orchestrators are a complicated topic.

**Reply by vitalykarasik** [reply]
> and it is not hardcoded to 1

I meant "either unlimited [by default, without 'concurrency'] or 1".

>GHA itself has no concept of throttling as you describe.

I see. But IMHO, it's an important feature that should be implemented in GHA.

Thank you again, Vitaly

**Reply by naikrovek** [reply]
It is implemented, just not as you want it; it's set by the number of runners you have available, and not by any user interface setting.   

What I would do, if it is not feasible to have long-lived runner VMs which take work as they are able, is to have whatever receives your webhook check to see how many runners are registered with your repo (or org).  There is an API call you can make to give you this information.  If the number of registered runners is less than the maximum number of runners that you've set, _then_ do the work that the webhook receiver currently does, and spawn a virtual machine.  If not, wait a reasonable amount of time, and then loop around try again, repeating that delay, check, delay, check loop until a slot is available.

I respect your desire to have this, and I understand why you want it.  GitHub Actions was just not designed with this configuration setting in mind.  To set the number of maximum jobs you want to run at a time, you make sure that you have no more than that number of runners connected and ready to work, either by manually setting up that many long-lived runners, or writing automation to manage short-lived runners for you.

Maybe this is something they will add in the future, but based on what little I know about how the Actions system works, I don't think it would be necessarily easy.

**Reply by vitalykarasik** [reply]
Thank you for the suggestions, I'll check them out.
Vitaly

**Reply by ViacheslavKudinov** [reply]
Hi,
In theory, you can use managed instance group on GCP with a limit of 3 instances and add your new VM to this group.
It is more complicated than just one sentence and needs some investigation, but you definitely need some orchestration of self-hosted runners for your use case.
Also, you can rise a feature request to GitHub to get the number of concurrent jobs.

**Reply by vitalykarasik** [reply]
@ViacheslavKudinov - thank you! Yes, I understand that DIY  orchestration is doable, I just hoped to find this feature in GHA or actions.
I'll open a feature request.
